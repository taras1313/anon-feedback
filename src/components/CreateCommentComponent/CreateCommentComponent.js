import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Chip from '@material-ui/core/Chip';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';

import cx from 'classnames';
import { uniqueNamesGenerator, adjectives, animals } from 'unique-names-generator';

const useStyles = makeStyles(theme => ({
  root: {
    width: '80%',
    height: '70vh'
  },
  chip: {
    marginLeft: '6px',
    height: '26px'
  },
  repliedToWrapper: {
    marginTop: '10px'
  },
  appBar: {
    position: 'relative',
    backgroundColor: '#424b5f',
    boxShadow: 'none'
  },
  title: {
    marginLeft: theme.spacing(2),
    textTransform: 'capitalize',
    flex: 1,
  },
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '30px 40px',
    flexDirection: 'column',
    flex: 1
  },
  textField: {
    width: '100%'
  },
  smallTextField: {
    width: 200
  },
  nickNameWrapper: {
    display: 'flex',
    flexDirection: 'column',
    marginTop: '10px'
  },
  formControl: {
    margin: 0
  },
  controlBtns: {
    marginTop: '16px',
    marginLeft: 'auto',
    textTransform: 'capitalize',
    display: 'flex',
    justifyContent: 'flex-end'
  },
  button: {
    fontWeight: 'bold',
    boxShadow: 'none',
    marginLeft: '16px'
  },
  submitBtn: {
    backgroundColor: '#65c178',
    fontWeight: 'bold',
    color: 'white',
    '&:hover': {
      backgroundColor: '#159776'
    },
  }
}));

export const CreateCommentComponent = (
  {
    username = '',
    onCreateComment,
    repliedTo,
    onRepliedToReset
  }
) => {
  const classes = useStyles();
  const [isChecked, setIsChecked] = React.useState(!!username);
  const [nickName, setNickName] = React.useState(username);
  const [commentText, setCommentText] = React.useState('');

  const toggleAutoGenerator = () => {
    const config = {
      dictionaries: [adjectives, animals],
      separator: ' ',
      length: 2,
    };
    
    const value = !isChecked ? uniqueNamesGenerator(config) : '';

    setNickName(value);
    setIsChecked(!isChecked);
  };

  const updateCommentText = ({ target: { value } }) => setCommentText(value);

  const updateUserNickName = ({ target: { value } }) => setNickName(value);

  const resetCommentFields = () => {
    setCommentText('');
    setIsChecked(false);
  };

  const createCommentHandler = () => {
    onRepliedToReset && onRepliedToReset();

    onCreateComment({ nickName, commentText, repliedTo })
      .then(resetCommentFields);
  };

  return (
    <div>
      {!username && (
        <div className={classes.nickNameWrapper}>
          <TextField
            autoComplete='off'
            id="standard-basic"
            className={cx(classes.textField, classes.smallTextField)}
            onChange={updateUserNickName}
            required
            value={nickName}
            name="username"
            label="Anon nick name"
          />
          <FormControlLabel
            className={classes.formControl}
            control={
              <Checkbox
                checked={isChecked}
                onChange={toggleAutoGenerator}
                color="primary"
              />
            }
            label="Generate nick name for you?"
          />
        </div>
      )}

      {repliedTo && (
        <div className={classes.repliedToWrapper}>
          Comment replied to
          <Chip
            label={`@${repliedTo}`}
            onDelete={onRepliedToReset}
            color="primary"
            variant="outlined"
            className={classes.chip}
          />
        </div>
      )}
      <TextField
        required
        autoComplete='off'
        id="standard-multiline-flexible"
        label="Your comment"
        multiline
        rows="6"
        value={commentText}
        onChange={updateCommentText}
        name="text"
        className={classes.textField}
        margin="normal"
        variant="filled" // ?
        // variant="outlined" // ?

      />

      <div className={classes.controlBtns}>
        <Button
          data-testid="here"
          variant="outlined"
          onClick={resetCommentFields}
          className={classes.button}
        >
          Reset
        </Button>

        <Button
          variant="outlined"
          onClick={createCommentHandler}
          color="primary"
          className={cx(classes.button)}
        >
          Create Comment
        </Button>
      </div>
    </div>
  );
};
