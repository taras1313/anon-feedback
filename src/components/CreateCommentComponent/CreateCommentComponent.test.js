import React from 'react';
import { shallow } from 'enzyme';
import { CreateCommentComponent } from './CreateCommentComponent';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { BrowserRouter as Router } from 'react-router-dom';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';

describe('Header component', () => {
	const setIsChecked = jest.fn();
	const setNickName = jest.fn();
	const onCreateComment = jest.fn().mockImplementation(() => ({ then: jest.fn() }));
	const setCommentText = jest.fn();
	const useState = jest.spyOn(React, 'useState');

	afterEach(() => jest.clearAllMocks());

	test('renders w/o props', () => {
		const wrapper = shallow(<CreateCommentComponent />);

		expect(wrapper).toMatchSnapshot();
	});

	test('Create Comment Component use state', () => {
		useState.mockImplementationOnce(() => [true, setIsChecked]);

		const wrapper = shallow(<CreateCommentComponent />);

		wrapper.find("[data-testid='here']").simulate('click');

		expect(setIsChecked).toHaveBeenCalled();
	});

	test('fake name', () => {
		useState.mockImplementationOnce(() => [false, setIsChecked]);
		useState.mockImplementationOnce(() => [true, setNickName]);

		const wrapper = shallow(<CreateCommentComponent />);

		wrapper.find(FormControlLabel).props().control.props.onChange();
		expect(setIsChecked).toHaveBeenCalledWith(true);
		expect(setNickName).toHaveBeenCalledTimes(1);
	});

	test('branches', () => {
		useState.mockImplementationOnce(() => [true, setIsChecked]);
		useState.mockImplementationOnce(() => [true, setNickName]);

		const wrapper = shallow(
			<CreateCommentComponent onRepliedToReset={jest.fn()} onCreateComment={onCreateComment} repliedTo />
		);

		wrapper.find(FormControlLabel).props().control.props.onChange();
		expect(setIsChecked).toHaveBeenCalledWith(false);
		expect(setNickName).toHaveBeenCalledTimes(1);

		wrapper.find(Button).at(1).props().onClick();
		expect(onCreateComment).toHaveBeenCalledTimes(1);
	});

	test('creates comment', () => {
		const wrapper = shallow(<CreateCommentComponent onCreateComment={onCreateComment} />);

		wrapper.find(Button).at(1).props().onClick();
		expect(onCreateComment).toHaveBeenCalledTimes(1);
	});

	test('updates comment and username', () => {
		useState.mockImplementationOnce(() => [true, setIsChecked]);
		useState.mockImplementationOnce(() => [true, setNickName]);
		useState.mockImplementationOnce(() => [true, setCommentText]);

		const wrapper = shallow(<CreateCommentComponent />);

		const event = { target: { value: 'test' } };

		wrapper.find(TextField).at(1).props().onChange(event);
		wrapper.find(TextField).at(0).props().onChange(event);

		expect(setCommentText).toHaveBeenCalledWith('test');
		expect(setNickName).toHaveBeenCalledWith('test');
	});
});
