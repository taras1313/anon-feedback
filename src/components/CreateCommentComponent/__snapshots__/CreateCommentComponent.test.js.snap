// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders w/o props 1`] = `
<div>
  <div
    className="makeStyles-nickNameWrapper-9"
  >
    <WithStyles(ForwardRef(TextField))
      autoComplete="off"
      className="makeStyles-textField-7 makeStyles-smallTextField-8"
      id="standard-basic"
      label="Anon nick name"
      name="username"
      onChange={[Function]}
      required={true}
      value=""
    />
    <WithStyles(ForwardRef(FormControlLabel))
      className="makeStyles-formControl-10"
      control={
        <ForwardRef(WithStyles)
          checked={false}
          color="primary"
          onChange={[Function]}
        />
      }
      label="Generate nick name for you?"
    />
  </div>
  <WithStyles(ForwardRef(TextField))
    autoComplete="off"
    className="makeStyles-textField-7"
    id="standard-multiline-flexible"
    label="Your comment"
    margin="normal"
    multiline={true}
    name="text"
    onChange={[Function]}
    required={true}
    rows="6"
    value=""
    variant="filled"
  />
  <div
    className="makeStyles-controlBtns-11"
  >
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-12"
      data-testid="here"
      onClick={[Function]}
      variant="outlined"
    >
      Reset
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-12"
      color="primary"
      onClick={[Function]}
      variant="outlined"
    >
      Create Comment
    </WithStyles(ForwardRef(Button))>
  </div>
</div>
`;
